zerver/webhooks/slack/view.py

```diff
@@ -1,62 +1,237 @@
-from zerver.lib.exceptions import JsonableError
-from zerver.lib.response import json_success
-from zerver.lib.typed_endpoint import typed_endpoint
+from zerver.lib.response import json_slack_challenge_response, json_success
+from zerver.lib.typed_endpoint import JsonBodyPayload, typed_endpoint
+from zerver.lib.validator import WildValue, check_none_or, check_string
 from zerver.lib.webhooks.common import check_send_webhook_message
 from zerver.models import UserProfile
 
+FILE_STR = "\n*[{file_name}]({file_link})*"
 ZULIP_MESSAGE_TEMPLATE = "**{message_sender}**: {text}"
 VALID_OPTIONS = {"SHOULD_NOT_BE_MAPPED": "0", "SHOULD_BE_MAPPED": "1"}
 
+SlackFileListT: TypeAlias = List[Dict[str, str]]
+SlackAPIResponseT: TypeAlias = Dict[str, Any]
+
+SLACK_CHANNELMENTION_REGEX = r"(?<=<#)(.*)(?=>)"
+
+
+def is_bot_message(payload: WildValue) -> bool:
+    app_api_id = payload.get("api_app_id").tame(check_none_or(check_string))
+    bot_app_id = (
+        payload.get("event", {})
+        .get("bot_profile", {})
+        .get("app_id")
+        .tame(check_none_or(check_string))
+    )
+    return bot_app_id is not None and app_api_id == bot_app_id
+
+
+def convert_mentions(text: str, app_token: str) -> str:
+    tokens = text.split(" ")
+    for iterator in range(len(tokens)):
+        slack_usermention_match = re.search(SLACK_USERMENTION_REGEX, tokens[iterator], re.VERBOSE)
+        slack_channelmention_match = re.search(
+            SLACK_CHANNELMENTION_REGEX, tokens[iterator], re.MULTILINE
+        )
+
+        if slack_usermention_match:
+            # We convert user mentions into a mention-like message to because
+            # there are no way of mapping between a users Slack and Zulip account
+
+            slack_id = slack_usermention_match.group(2)
+            user_obj = get_slack_user_info(user_id=slack_id, token=app_token)
+            tokens[iterator] = "@**" + user_obj["name"] + "**"
+
+        elif slack_channelmention_match:
+            # We convert channel mentions to a mention-like messages to avoid
+            # mentioning the wrong channel over at Zulip.
+
+            channel_info: List[str] = slack_channelmention_match.group(0).split("|")
+            channel_name = channel_info[1]
+            tokens[iterator] = f"**#{channel_name}**" if channel_name else "**#[private channel]**"
+
+            # TODO: as a follow up, we could probably discuss and decide on a
+            # more proper way to handle messages mentioning channels.
+
+    text = " ".join(tokens)
+    return text
+
+
+def get_slack_user_info(user_id: str, token: str) -> SlackAPIResponseT:
+    # TODO: additional improvement would be to cache the response
+    user_obj = get_slack_api_data(
+        "https://slack.com/api/users.info", get_param="user", token=token, user=user_id
+    )
+    return user_obj
+
+
+def get_slack_channel_info(channel_id: str, token: str) -> SlackAPIResponseT:
+    # TODO: additional improvement would be to cache the response
```

Comment:
FOLLOW-UP: This integration expects high-volume traffic because we're sending live messages from Slack to Zulip. Doing two callbacks(username & channel) for each message could potentially be problematic for latency especially if this integration is used as part of the Slack Bridge, which is trying to replicate a real-time mirror to Slack. Adding caching might ease the potential latency problem.